### DISCUSSION TITLE
Result of GetInternalForceAt()

### DISCUSSION THREAD

**Discussion by mauromen7** [BriefFiniteElementNet/BriefFiniteElement.Net#123] [Q&A] [answered]
Hello, I have known the library for several years and I see with pleasure that they continue to work in the library for this I congratulate and thank them.
Several years ago I started an application for analysis of continuous beams which I forgot for other obligations, now I would like to resume it and update it with the latest version of the library and I find some updates in syntax that I must correct.
I wrote a small routine to analyze a simple beam and I do not see agreement between the results of the internal forces and the reactions, additionally the internal forces evaluated at different stations do not result in the values that I expect.

```
var model = new Model();
            var sec = new Sections.UniformGeometric1DSection(SectionGenerator.GetISetion(0.24, 0.67, 0.01, 0.006));
            var mat = Materials.UniformIsotropicMaterial.CreateFromYoungPoisson(210000000000.0, 0.3);
            model.Nodes.Add(new Node(0, 0, 0) { Label = "n0" });
            model.Nodes.Add(new Node(4, 0, 0) { Label = "n1" });
                     
            model.Elements.Add(new BarElement(model.Nodes["n0"], model.Nodes["n1"]) { Label = "e0" });

            model.Nodes["n0"].Constraints = Constraints.MovementFixed & Constraints.FixedRX;
            model.Nodes["n1"].Constraints = Constraints.MovementFixed & Constraints.FixedRX;

            (model.Elements["e0"] as BarElement).Material = mat;
            (model.Elements["e0"] as BarElement).Section = sec;

            var u1 = new Loads.UniformLoad(LoadCase.DefaultLoadCase, -Vector.K, -1000, CoordinationSystem.Global);
            model.Elements["e0"].Loads.Add(u1);

            model.Solve_MPC();
            
            var n0Force = model.Nodes["n0"].GetSupportReaction();
            Console.WriteLine("support reaction of n0: {0}", n0Force);

            var elm = model.Elements[0] as BarElement;
            var frc = elm.GetInternalForceAt(0);
            Console.WriteLine("I.F. of in 0.0*L: {0}", frc);
            var frc2 = elm.GetInternalForceAt(0.5);
            Console.WriteLine("I.F. of in 0.5*L: {0}", frc2);
            var frc3 = elm.GetInternalForceAt(1.0);
            Console.WriteLine("I.F. of in 1.0*L: {0}", frc3);

            var n1Force = model.Nodes["n1"].GetSupportReaction();
            Console.WriteLine("support reaction of n1: {0}", n1Force);

            Console.ReadLine();
```

Ra <> Fz(0) **'this is error**, 
evaluated as:
model.Nodes["n0"].GetSupportReaction() <> elm1.GetInternalForceAt (0)  **'this is error**, 

![211117_8731](https://user-images.githubusercontent.com/23561312/142346848-bfdcffff-b9ba-435d-a683-976c935c4c0d.png)

Several years ago a similar error was found but it was corrected, so it is likely that there is an error in my model definition routine.
I would appreciate your comments, Could you tell me what is the problem in the definition of the code
I remain attentive

**Comment by epsi1on** [comment, marked as answer]
I think using `BarElement.GetExactInternalForce()` instead of `BarElement.GeInternalForce()` will give you the result that you expect.

[Here](https://bfenet.readthedocs.io/en/latest/elements/finiteElements/Bar/internalforces.html#barelement-internalforce) is some more info on difference about  `BarElement.GetExactInternalForce()` vs `BarElement.GeInternalForce()` .

**Reply by mauromen7** [reply]
thanks for your attention, considering the function "GetExactInternalForceAt ()" the expected results are obtained in GetExactInternalForceAt (-0.999) for start, GetExactInternalForceAt (0) for middle and  GetExactInternalForceAt (0.999) for end

the example indicated is: 

```
  ______________________  q
 |                       |
 |          e1           |
#========================#
n1                       n2

               public static void Run2()
        {
            var model = new Model();

            Node n1, n2;

            model.Nodes.Add(n1 = new Node(0, 0, 0) { Constraints = Constraints.Fixed });
            model.Nodes.Add(n2 = new Node(1, 0, 0) { Constraints = Constraints.Fixed });

            var elm = new BarElement(n1, n2);

            elm.Section = new BriefFiniteElementNet.Sections.UniformParametric1DSection(a: 0.01, iy: 0.01, iz: 0.01, j: 0.01);
            elm.Material = BriefFiniteElementNet.Materials.UniformIsotropicMaterial.CreateFromYoungPoisson(210e9, 0.3);

            var load = new Loads.UniformLoad();

            load.Case = LoadCase.DefaultLoadCase;
            load.CoordinationSystem = CoordinationSystem.Global;
            load.Direction = Vector.K;
            load.Magnitude = 10;

            elm.Loads.Add(load);
            model.Elements.Add(elm);

            model.Solve_MPC();

            //var f1 = elm.GetInternalForceAt(0);
            //var f2 = elm.GetExactInternalForceAt(0);

            Console.WriteLine("##### evaluate support reaction #####");
            var n0Force = model.Nodes[0].GetSupportReaction();
            Console.WriteLine("support reaction of n0: {0}", n0Force);
            var n1Force = model.Nodes[1].GetSupportReaction();
            Console.WriteLine("support reaction of n1: {0}", n1Force);

            Console.WriteLine("");
            Console.WriteLine("##### evaluate whith GetInternalForceAt() #####");
            var f1 = elm.GetInternalForceAt(0);
            Console.WriteLine("I.F. of in 0.0*L: {0}", f1);

            Console.WriteLine("");
            Console.WriteLine("##### evaluate whith GetExactInternalForceAt() #####");
            var f1e = elm.GetExactInternalForceAt(-0.9999);
            Console.WriteLine("I.F.E. of in 0.0*L: {0}", f1e);

            var f2e = elm.GetExactInternalForceAt(0);
            Console.WriteLine("I.F.E. of in 0.5*L: {0}", f2e);
            
            Force f3e = elm.GetExactInternalForceAt(0.9999);
            Console.WriteLine("I.F.E. of in 1.0*L: {0}", f3e);

            Console.ReadLine();
```
//##### evaluate support reaction #####
//support reaction of n0: F: 0, 0, -5, M: 0, 0.833333333333333, 0
//support reaction of n1: F: 0, 0, -5, M: 0, -0.833333333333333, 0

//##### evaluate whith GetInternalForceAt() #####
//I.F.of in 0.0 * L: F: 0, 0, 0, M: 0, 0, 0

//##### evaluate whith GetExactInternalForceAt() #####
//I.F.E.of in 0.0 * L: F: 0, 0, 4.9995, M: 0, -0.833083345833333, 0
//I.F.E.of in 0.5 * L: F: 0, 0, 0, M: 0, 0.416666666666667, 0
//I.F.E.of in 1.0 * L: F: 0, 0, -4.9995, M: 0, -0.833083345833334, 0

Additionally, write the following model and the results are consistent with those previously presented.

```
  ______________________  q
 |                       |
 |      e1        e2     |
#===========<>===========#
n1          n2             n3

       public static void Run3()
        {
            var model = new Model();

            Node n1, n2, n3;

            model.Nodes.Add(n1 = new Node(0, 0, 0) { Constraints = Constraints.Fixed });
            model.Nodes.Add(n2 = new Node(0.5, 0, 0) { Constraints = Constraints.RotationFixed });
            model.Nodes.Add(n3 = new Node(1, 0, 0) { Constraints = Constraints.Fixed });

            var elm1 = new BarElement(n1, n2);
            var elm2 = new BarElement(n2, n3);

            elm1.Section = new BriefFiniteElementNet.Sections.UniformParametric1DSection(a: 0.01, iy: 0.01, iz: 0.01, j: 0.01);
            elm1.Material = BriefFiniteElementNet.Materials.UniformIsotropicMaterial.CreateFromYoungPoisson(210e9, 0.3);

            elm2.Section = new BriefFiniteElementNet.Sections.UniformParametric1DSection(a: 0.01, iy: 0.01, iz: 0.01, j: 0.01);
            elm2.Material = BriefFiniteElementNet.Materials.UniformIsotropicMaterial.CreateFromYoungPoisson(210e9, 0.3);

            var load = new Loads.UniformLoad();

            load.Case = LoadCase.DefaultLoadCase;
            load.CoordinationSystem = CoordinationSystem.Global;
            load.Direction = Vector.K;
            load.Magnitude = 10;

            elm1.Loads.Add(load);
            model.Elements.Add(elm1);

            elm2.Loads.Add(load);
            model.Elements.Add(elm2);

            model.Solve_MPC();

            Console.WriteLine("##### evaluate support reaction #####");
            var n0Force = model.Nodes[0].GetSupportReaction();
            Console.WriteLine("support reaction of n0: {0}", n0Force);
            var n1Force = model.Nodes[2].GetSupportReaction();
            Console.WriteLine("support reaction of n2: {0}", n1Force);

            Console.WriteLine("");
            Console.WriteLine("##### evaluate whith GetInternalForceAt() #####");
            var f1i = elm1.GetInternalForceAt(-1);
            Console.WriteLine("I.F. of in 0.0*L: {0}", f1i);
            var f1j = elm1.GetInternalForceAt(1);
            Console.WriteLine("I.F. of in 0.5*L: {0}", f1j);
            var f2i = elm2.GetInternalForceAt(-1);
            Console.WriteLine("I.F. of in 0.5*L: {0}", f2i);
            var f2j = elm2.GetInternalForceAt(1);
            Console.WriteLine("I.F. of in 1.0*L: {0}", f2j);

            Console.WriteLine("");
            Console.WriteLine("##### evaluate whith GetExactInternalForceAt() #####");

            var f1e = elm1.GetExactInternalForceAt(-0.99999);
            Console.WriteLine("I.F.E. of in 0.0*L: {0}", f1e);

            //var f2e = elm1.GetExactInternalForceAt(1); //Error in barElemnt.cs line 982
            var f2e = elm1.GetExactInternalForceAt(0.999999);
            Console.WriteLine("I.F.E. of in 0.5*L: {0}", f2e);

             f2e = elm2.GetExactInternalForceAt(-0.9999);
            Console.WriteLine("I.F.E. of in 0.5*L: {0}", f2e);

            // var f3e = elm.GetExactInternalForceAt(1); //Error in barElemnt.cs line 982
            Force f3e = elm2.GetExactInternalForceAt(0.9999);
            Console.WriteLine("I.F.E. of in 1.0*L: {0}", f3e);

            Console.ReadLine();
```
//##### evaluate support reaction #####
//support reaction of n0: F: 0, 0, -5, M: 0, 0.833333333333333, 0
//support reaction of n2: F: 0, 0, -5, M: 0, -0.833333333333333, 0

//##### evaluate whith GetInternalForceAt() #####
//I.F.of in 0.0 * L: F: 0, 0, 2.5, M: 0, -0.625, 0
//I.F.of in 0.5 * L: F: 0, 0, 2.5, M: 0, 0.625, 0
//I.F.of in 0.5 * L: F: 0, 0, -2.5, M: 0, 0.625, 0
//I.F.of in 1.0 * L: F: 0, 0, -2.5, M: 0, -0.625, 0
// the answer will be more approximate if a greater number of nodes is generated

//##### evaluate whith GetExactInternalForceAt() #####
//I.F.E.of in 0.0 * L: F: 0, 0, 4.999975, M: 0, -0.833320833364584, 0
//I.F.E.of in 0.5 * L: F: 0, 0, 2.50000000168171E-06, M: 0, 0.416666666666355, 0
//I.F.E.of in 0.5 * L: F: 0, 0, -0.00025000000000075, M: 0, 0.416666663541667, 0
 //I.F.E.of in 1.0 * L: F: 0, 0, -4.99975, M: 0, -0.833208336458334, 0

Thanks for your help, solved my problem. I hope these examples are useful for another user

**Reply by epsi1on** [reply]
Start node and end node (e.g `xi =-1.0` and `xi=1.0`) are special locations where theoretically internal force are not defined as a specific value, so this library will throw exception if you try to pass those values to `BarElement.GetExactInternalForceAt()` and `BarElement.GetExactInternalForceAt()`. but a little after that, say `xi=1-1e-9=0.999999999999` is not special like that. here is the code that throws exception:

https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/blob/983677507ed606284bd95182ab48579e73146a8a/BriefFiniteElementNet/Elements/BarElement.cs#L989

another example is where there is a concentrated load on the beam:

![image](https://user-images.githubusercontent.com/7839141/143258692-32903c24-5e02-482d-934c-7574ffcd4cef.png)


The shear force is not a single value where the load is applied, but the moment is single value. but currently code do throws exception.
